src/query/parser/promql/parse.go

```diff
@@ -405,3 +415,25 @@ func (p *parseState) walk(node pql.Node) error {
 		return fmt.Errorf("promql.Walk: unhandled node type %T, %v", node, node)
 	}
 }
+
+func resolveStringArgument(expr pql.Expr) (string, error) {
+	switch e := expr.(type) {
+	case *pql.StringLiteral:
+		return e.Val, nil
+	case *pql.ParenExpr:
+		return resolveStringArgument(e.Expr)
+	default:
+		return "", fmt.Errorf("promql.resolveStringArgument: unexpected expression type %T", e)
+	}
+}
+
+func resolveMatrixRangeArgument(expr pql.Expr) (time.Duration, error) {
+	switch e := expr.(type) {
+	case *pql.MatrixSelector:
+		return e.Range, nil
+	case *pql.ParenExpr:
+		return resolveMatrixRangeArgument(e.Expr)
+	default:
+		return time.Duration(0), fmt.Errorf("promql.resolveMatrixRangeArgument: unexpected expression type %T", e)
+	}
+}
```

Comment:
I'd suggest unwrapping each `*pql.Call` argument before use, similar to how prom does it: https://github.com/prometheus/prometheus/commit/3a204be6b75273c28c36ea4934239e4348b81609#diff-ad2761bb162a2f12326b18f2714ef54f4cd476cfe3388f77b2b20c5ff82e4faaR1062

You don't have to worry about additional errors from the unwrapped expressions being the wrong type since it wouldn't have passed the parsing step otherwise

BTW you may want to coordinate with Linas since it may get rolled into this PR: https://github.com/m3db/m3/pull/2805

Reply:
https://github.com/m3db/m3/pull/2809/commits/597a4783106c023e2ac60153ccabfa36cb5f1b41#diff-4d8dd30f3d10b12a442c2abb2490e729062172a5eff850dbe38a7bf083db71feL294-R309